src/Lykke.Service.CryptoIndex.DomainServices/LCI10/LCI10Calculator.cs

```diff
@@ -0,0 +1,300 @@
+            var notFoundAssets = assets.Where(x => !wantedMarketCaps.Select(mc => mc.Asset).Contains(x)).ToArray();
+            if (wantedMarketCaps.Count() != assets.Count)
+                throw new InvalidOperationException($"Can't find assets from settings in CoinMarketCap data: {notFoundAssets.ToJson()}.");
+
+            var totalMarketCap = wantedMarketCaps.Select(x => x.MarketCap.Value).Sum();
+
+            lock (_sync)
+            {
+                _marketCaps.Clear();
+                _marketCaps.AddRange(wantedMarketCaps);
+                _weights.Clear();
+                foreach (var asset in assets)
+                {
+                    var assetMarketCap = wantedMarketCaps.Single(x => x.Asset == asset).MarketCap.Value;
+                    var assetWeight = assetMarketCap / totalMarketCap;
+                    _weights[asset] = assetWeight;
+                }
+            }
+
+            _log.Info($"Finished calculating weights for {assets.Count} assets.");
+        }
+
+        private async Task CalculateIndex(ITimerTrigger timer, TimerTriggeredHandlerArgs args, CancellationToken ct)
+        {
+            try
+            {
+                await CalculateIndex();
+            }
+            catch (Exception e)
+            {
+                _log.Error(e);
+            }
+        }
+
+        private async Task CalculateIndex()
+        {
+            _log.Info("Started calculating index...");
+
+            var settings = await _settingsService.GetAsync();
+
+            List<AssetMarketCap> marketCaps;
+            IDictionary<string, decimal> weights;
+            var assetsPrices = await _tickPricesService.GetPricesAsync();
+            var lastIndex = await _indexStateRepository.GetAsync();
+
+            RecalculateTheWeightsIfSomeWeightsAreNotFound(settings.Assets);
+
+            lock (_sync)
+            {
+                marketCaps = _marketCaps.ToList();
+                weights = _weights.Clone();
+            }
+
+            if (!IsAllDataArePresented(marketCaps, weights, assetsPrices))
+            {
+                _log.Info("Skipped LCI10 calculation.");
+                return;
+            }
+
+            // Save index as previous for the next execution
+            var indexState = CalculateIndexState(settings.Assets, weights, assetsPrices, lastIndex);
+            await _indexStateRepository.SetAsync(indexState);
+
+            // Save index to history
+            var indexHistory = new IndexHistory(indexState.Value, marketCaps, weights, indexState.MiddlePrices, DateTime.UtcNow);
+            await _indexHistoryRepository.InsertAsync(indexHistory);
+
+            // Publish index to RabbitMq
+            var tickPrice = new Domain.TickPrice.TickPrice(Lci10, Lci10, indexHistory.Value, indexHistory.Value, indexHistory.Time);
+            _tickPricePublisher.Publish(tickPrice);
+
+            _log.Info($"Finished calculating index for {settings.Assets.Count} assets, value: {indexState.Value}.");
```

Comment:
Variables can be put to the context

Reply:
Prefer two simple step instead of one complex. It looks more readable to me.